### PR TITLE
Add Qwen 3.5 model to hub

### PR DESCRIPTION
**Pull request by laxmareddyp**
## Description of the change

This PR address the feature request #2612 issue.

 ## Qwen3.5 Model:

* Represents a departure from traditional Transformer-only designs, moving toward a hybrid **"native multimodal"** architecture.  
* introduces a significant architectural shift from the previous generation by moving toward a **native multimodal agent** foundation. 
* The core innovation is a hybrid design that combines **Gated Delta Networks (GDN)** with sparse **Mixture-of-Experts (MoE)**, specifically optimized for high-throughput inference and massive context windows up to 1 million tokens.
* It is designed to function as an **"all-in-one AI worker"** capable of seeing, reading, coding, and planning within a single unified backbone.

## Core Architectural Features:

* **Hybrid Attention Core:** 

    * To overcome the quadratic scaling limitations of standard attention, Qwen 3.5 uses a 3:1 hybrid layout. It alternates three Gated DeltaNet layers (linear attention) with one standard Gated Attention layer (full softmax attention).
     *  **Efficiency:** GDN maintains a fixed-size state matrix, keeping compute and memory costs linear relative to sequence length.
    * **Structure:** The model typically alternates between GDN layers and standard full attention layers in a 3:1 ratio (e.g., three GDN blocks followed by one full attention block).

* **Sparse Mixture-of-Experts (MoE):** The flagship models utilize high-sparsity routing to maintain massive knowledge capacity with low compute costs.

* **Native Multimodality (Early Fusion):** Unlike previous versions that used "late fusion" (attaching a separate vision encoder), Qwen3.5 uses early fusion. Text and visual tokens enter the same backbone from the start, allowing the model to process mixed content—like screenshots and code—simultaneously.

Numerical Verifications:

Screenshot 2026-04-09 at 10 07 39 AM
Screenshot 2026-04-09 at 10 14 42 AM

Screenshot 2026-04-06 at 11 31 31 PM

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/qwen3_5/qwen3_5_causal_lm_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+                    current_pos += 1
+                    i += 1
+
+            # Channel layout: [text, temporal, height, width].
+            all_pos[b, 0] = t_pos
+            all_pos[b, 1] = t_pos
+            all_pos[b, 2] = h_pos
+            all_pos[b, 3] = w_pos
+
+        return tf.constant(all_pos, dtype="int32")
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5/qwen3_5_causal_lm_preprocessor.py:453]: No. This is the standard KerasHub pattern:

`_compute_position_ids `runs inside generate_preprocess which is decorated with @preprocessing_function — it always runs eagerly on CPU, never inside tf.function or JIT
M-RoPE position ID computation is inherently sequential (you need to know where each vision span ends before assigning the next position) — vectorizing it would be extremely complex for no benefit
Gemma3's preprocessor does the same thing — Python loops + numpy for position ID computation.

The gemini comment is technically correct in isolation but ignores the architectural context. Will not refactor.

File: keras_hub/src/models/qwen3_5/qwen3_5_gated_delta_net.py [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+        correction = ops.squeeze(correction, axis=-2)
+        new_row = attn[..., i, :]
+        update = ops.pad(
+            correction, [[0, 0], [0, 0], [0, 0], [0, chunk_size - i]]
+        )
+        new_row = new_row + update
+        before = attn[..., :i, :]
+        after = attn[..., i + 1 :, :]
+        new_row_exp = ops.expand_dims(new_row, axis=-2)
+        attn = ops.concatenate([before, new_row_exp, after], axis=-2)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5/qwen3_5_gated_delta_net.py:157]: The iterative Neumann-series correction loop is a faithful port of HF's `torch_chunk_gated_delta_rule ` which uses the same in-place loop. It's O(chunk_size) = O(64) iterations — not O(seq_len). Using ops.slice_update would tie us to specific backends (it's not well-supported across TF/JAX/PyTorch), and the concatenate approach works correctly and matches HF's reference implementation. optimizing this is not worth the risk of introducing numerical bugs in the delta rule computation.

File: keras_hub/src/models/qwen3_5/qwen3_5_vision_encoder.py [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+            )  # (T, H*W, H)
+            ms = self.spatial_merge_size
+            patch_embed = ops.reshape(
+                patch_embed, (t, h // ms, ms, w // ms, ms, self.hidden_size)
+            )
+            patch_embed = ops.transpose(patch_embed, (0, 1, 3, 2, 4, 5))
+            patch_embed = ops.reshape(patch_embed, (-1, self.hidden_size))
+            all_embeds.append(patch_embed)
+
+        return ops.concatenate(all_embeds, axis=0)
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply keras_hub/src/models/qwen3_5/qwen3_5_vision_encoder.py:629]: The vision encoder's call() always receives concrete (eager) grid_thw values — it's called from generate_step and score with real data, never inside a tf.function trace. The int() calls are safe because grid_thw is always a concrete tensor at this point. HF's reference implementation uses the same pattern. Vectorizing this would be a major refactor with high risk of introducing numerical bugs in the bilinear interpolation, for no practical benefit.

File: tools/checkpoint_conversion/convert_qwen3_5_checkpoints.py [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+    print("TEXT-ONLY VALIDATION")
+    print("=" * 50)
+
+    hf_ids = hf_results["text_token_ids"]
+
+    # --- Token ID parity ---
+    keras_preprocessed = keras_model.preprocessor.generate_preprocess(
+        TEXT_PROMPT, sequence_length=hf_ids.shape[1]
+    )
+    keras_ids = ops.convert_to_numpy(keras_preprocessed["token_ids"])
```
Comment (sachinprasadhs) [inline tools/checkpoint_conversion/convert_qwen3_5_checkpoints.py:336]: How much difference is expected, based on calculation can we record this and minus.

File: keras_hub/src/models/qwen3_5/qwen3_5_video_converter.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+from keras import ops
+
+from keras_hub.src.api_export import keras_hub_export
+from keras_hub.src.models.qwen3_5.qwen3_5_backbone import Qwen3_5Backbone
+from keras_hub.src.utils.tensor_utils import in_tf_function
+from keras_hub.src.utils.tensor_utils import preprocessing_function
+
+
+@keras_hub_export("keras_hub.models.Qwen3_5VideoConverter")
+class Qwen3_5VideoConverter(keras.layers.Layer):
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_video_converter.py]: You need to create a new VideoConverter base class and subclass it here.

File: keras_hub/src/models/qwen3_5/qwen3_5_video_converter.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+        self.temporal_patch_size = temporal_patch_size
+        self.spatial_merge_size = spatial_merge_size
+        self.min_pixels = min_pixels
+        self.max_pixels = max_pixels
+        self.image_mean = image_mean or [0.5, 0.5, 0.5]
+        self.image_std = image_std or [0.5, 0.5, 0.5]
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_video_converter.py]: Does it hold good for all the cases?

File: keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.patch_size = patch_size
+        self.temporal_patch_size = temporal_patch_size
+        self.spatial_merge_size = spatial_merge_size
+        self.min_pixels = min_pixels
+        self.max_pixels = max_pixels
+        self.image_mean = image_mean or [0.5, 0.5, 0.5]
+        self.image_std = image_std or [0.5, 0.5, 0.5]
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py]: We can pull this from config.

File: keras_hub/src/models/qwen3_5/qwen3_5_causal_lm_preprocessor.py [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+        result["vision_indices"] = vision_indices
+        result["position_ids"] = pos_ids if batched else tf.squeeze(pos_ids, 0)
+        return result
+
+    def _preprocess_images(self, images, batched):
+        """Convert raw images to patch tensors using the image converter.
+
+        Args:
+            images: A single PIL/numpy image, a list of images, or a
+                batched tensor.
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_causal_lm_preprocessor.py]: Remove PIL

File: keras_hub/src/models/qwen3_5/qwen3_5_causal_lm_preprocessor.py [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+    # Special tokens that the KerasHub BPE tokenizer may not encode
+    # as single tokens. Map from string → token ID.
+    SPECIAL_TOKEN_MAP = {
+        "<|im_start|>": 248045,
+        "<|im_end|>": 248044,
+        "<|vision_start|>": 248053,
+        "<|vision_end|>": 248054,
+        "<|image_pad|>": 248056,
+        "<|video_pad|>": 248057,
+    }
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_causal_lm_preprocessor.py]: I don't think we handle special tokens this way by mapping to token ID.

File: tools/checkpoint_conversion/convert_qwen3_5_checkpoints.py (hunk 4 above)
Comment (laxmareddyp) [reply tools/checkpoint_conversion/convert_qwen3_5_checkpoints.py:336]: No need to calculate now because no of parameters matches , it's dead code. removed in latest commit.

File: keras_hub/src/utils/transformers/convert_qwen3_5.py [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+    """Return kwargs for Qwen3_5VideoConverter, or None for text-only.
+
+    Uses the same vision config as images since HF's ``get_video_features``
+    calls ``get_image_features`` — identical encoder, identical parameters.
+    """
+    if "vision_config" not in transformers_config:
+        return None
+
+    vision_config = transformers_config["vision_config"]
+    preprocessor_config = load_json(preset, "preprocessor_config.json")
```
Comment (sachinprasadhs) [inline keras_hub/src/utils/transformers/convert_qwen3_5.py]: This should be video preprocessor config.

File: keras_hub/src/utils/transformers/convert_qwen3_5.py [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+    # HF preprocessor_config nests pixel limits under "size".
+    size_config = preprocessor_config["size"]
+    return {
+        "patch_size": vision_config["patch_size"],
+        "temporal_patch_size": vision_config["temporal_patch_size"],
+        "spatial_merge_size": vision_config["spatial_merge_size"],
+        "min_pixels": size_config["shortest_edge"],
+        "max_pixels": size_config["longest_edge"],
+        "image_mean": preprocessor_config["image_mean"],
+        "image_std": preprocessor_config["image_std"],
```
Comment (sachinprasadhs) [inline keras_hub/src/utils/transformers/convert_qwen3_5.py]: This should be scale and offset, we don't save mean and std anymore.

File: keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+
+    def __init__(
+        self,
+        patch_size=16,
+        temporal_patch_size=2,
+        spatial_merge_size=2,
+        min_pixels=65536,
+        max_pixels=16777216,
+        image_mean=None,
+        image_std=None,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py]: This should be scale and offset, we don't save mean and std anymore.

File: keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+            (target_h, target_w),
+            method=tf.image.ResizeMethod.BICUBIC,
+            antialias=True,
+        )[0]
+        image = tf.clip_by_value(image, 0.0, 255.0)
+
+        # Normalise to [-1, 1].
+        mean = tf.constant(self.image_mean, dtype="float32") * 255.0
+        std = tf.constant(self.image_std, dtype="float32") * 255.0
+        image = (image - mean) / std  # (H, W, 3)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py]: This should be scale and offset, we don't save mean and std anymore.
This needs to come from converter file.

File: keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+            size=(target_h, target_w),
+            interpolation="bicubic",
+            antialias=True,
+        )[0]
+        image = ops.clip(image, 0.0, 255.0)
+
+        # Normalise to [-1, 1].
+        mean = ops.array(self.image_mean, dtype="float32") * 255.0
+        std = ops.array(self.image_std, dtype="float32") * 255.0
+        image = (image - mean) / std  # (H, W, 3)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/qwen3_5/qwen3_5_image_converter.py]: This should be scale and offset, we don't save mean and std anymore.
